service/workspace/utilities.go

```diff
@@ -70,3 +73,154 @@ func (a *WorkspaceAPI) RecursiveList(ctx context.Context, path string) ([]Object
+// In addition, if the `path` is imported as a notebook, then the `item`’s
+// extension is automatically removed.
+//
+// workspace.UploadLanguage(...) is only required if source format.
+func (a *WorkspaceAPI) Upload(ctx context.Context, path string, r io.Reader, opts ...UploadOption) error {
+	buf := &bytes.Buffer{}
+	w := multipart.NewWriter(buf)
+	err := w.WriteField("path", path)
+	if err != nil {
+		return fmt.Errorf("write path: %w", err)
+	}
+	content, err := w.CreateFormFile("content", "content")
+	if err != nil {
+		return fmt.Errorf("write content: %w", err)
+	}
+	_, err = io.Copy(content, r)
+	if err != nil {
+		return fmt.Errorf("copy io: %w", err)
+	}
+	i := &Import{}
+	for _, v := range opts {
+		v(i)
+	}
+	if i.Format == "" || i.Format == ExportFormatSource {
+		for sfx, lang := range map[string]Language{
+			".py":    LanguagePython,
+			".sql":   LanguageSql,
+			".scala": LanguageScala,
+			".R":     LanguageR,
+		} {
+			if !strings.HasSuffix(path, sfx) {
+				continue
+			}
+			i.Language = lang
+		}
+	}
+	if i.Format != "" {
+		err = w.WriteField("format", i.Format.String())
+		if err != nil {
+			return fmt.Errorf("write format: %w", err)
+		}
+	}
+	if i.Language != "" {
+		err = w.WriteField("language", i.Language.String())
+		if err != nil {
+			return fmt.Errorf("write language: %w", err)
+		}
+	}
+	if i.Overwrite {
+		err = w.WriteField("overwrite", "true")
+		if err != nil {
+			return fmt.Errorf("write overwrite: %w", err)
+		}
+	}
+	err = w.Close()
+	if err != nil {
+		return fmt.Errorf("write close: %w", err)
+	}
+	impl, ok := a.impl.(*workspaceImpl)
+	if !ok {
+		return fmt.Errorf("wrong impl: %v", a.impl)
+	}
+	return impl.client.Do(ctx, "POST", "/api/2.0/workspace/import", buf.Bytes(), nil,
+		func(r *http.Request) error {
+			r.Header.Set("Content-Type", w.FormDataContentType())
+			return nil
+		})
+}
+
+// WriteFile is identical to [os.WriteFile] but for Workspace File.
+// Keep in mind: It doesn't upload the notebook, but the file and does
+// always overwrite it.
+func (a *WorkspaceAPI) WriteFile(ctx context.Context, name string, data []byte) error {
+	return a.Upload(ctx, name, bytes.NewBuffer(data),
+		UploadFormat(ExportFormatAuto),
+		UploadOverwrite())
+}
+
+type DownloadOption = func(q map[string]any)
+
+func DownloadFormat(f ExportFormat) func(q map[string]any) {
+	return func(q map[string]any) {
+		q["format"] = f.String()
+	}
+}
+
+// Download a notebook or file from the workspace by path.
+//
+// By default, it acts as if workspace.DownloadFormat(workspace.ExportFormatSource) option is supplied. When using
+// workspace.ExportFormatAuto, the `path` is imported or exported as either a workspace file or a notebook, depending
+// on an analysis of the `item`’s extension and the header content provided in the request.
+//
+// Returns [bytes.Buffer] of the path contents.
+func (a *WorkspaceAPI) Download(ctx context.Context, path string, opts ...DownloadOption) (*bytes.Buffer, error) {
```

Comment:
Please make this return an `io.ReadCloser` as well for symmetry with the Files API.